Ability to advanced search using multiple filters as "or"/"any" condition

mddilley opened the pull request:
## Associated issues
https://github.com/cityofaustin/atd-data-tech/issues/11282

This PR:
- adds an advanced search radio group to select either **and** or **or** logic
- updates `useAdvancedSearch` to handle and apply that logic to the advanced filters in the GraphQL query
- updates the code to use React Router [useSearchParams](https://reactrouter.com/en/main/hooks/use-search-params#usesearchparams) to manage the url query strings

## Testing
**URL to test:** 
<!---
 [branch-name]--atd-moped-main.netlify.app/moped/ if test instance or deploy-preview-[pr-number]--atd-moped-main.netlify.app/moped/ 
--->
https://deploy-preview-1187--atd-moped-main.netlify.app/moped/

**Steps to test:**
1. Go to the project list view in the deploy preview
2. Add an advanced filter by click the filter icon in the search box
3. Try searching for:
- **ALL**
- **Lead is COA ATD Project Delivery**
- **Sponsor is COA ATD Active Transportation & Street Design**
- Notice that there are radios for **all** and **any** above the advanced filter rows. Keep it as **all**
- You should see one result that created that is named **#11282 test project** that matches both of the filters
- Check the url and you should see two query parameters: **filter** (with the encoded filter string) and **isOr** which should have the value of **false**
4. Now, try searching for:
- **ANY**
- **Lead is COA ATD Project Delivery**
- **Sponsor is COA ATD Active Transportation & Street Design**
- Change the radios to match **any**
- You should see many results that meet either of the two filters
- Check the url and you should see two query parameters: **filter** (with the encoded filter string) and **isOr** which should have the value of **true**
5. Try combinations of advanced filters like **projects where Arterial Management is either the Lead, Sponsor, or a Partner** from the issue, and try to break it! (You might run into https://github.com/cityofaustin/atd-data-tech/issues/14495 - that fix is coming in the next PR)
6. Try copying your url with **filter** and **isOr** parameters into a different tab and the UI should reflect the advanced filters stored in the encoded string with the correct any/all option selected in the radio buttons
7. Try applying advanced filters to the view and then entering a search string into the search box and pressing enter. You should see the advanced filter UI reset and the query parameters removed.

---
#### Ship list
- [ ] Code reviewed 
- [ ] Product manager approved
- [ ] Added to [QA test script](https://docs.google.com/spreadsheets/d/1n_O6MLh9cwwPf57HUM394Ea-z9uuoEV1-QW4axNZXLE/edit#) if applicable

mddilley commented:
@johnclary Request you first to check in on the initial design

johnclary commented:
@mddilley just realized i never circled back on the toggle design. let's go with the radio buttons instead of the button toggle. 
![radio (1)](https://github.com/cityofaustin/atd-moped/assets/14793120/79341368-ac60-4dfc-8478-6f66f1267e0e)

could you hide these until a second filter is added?

mddilley commented on moped-editor/src/components/GridTable/Filters.js:
```diff
@@ -435,10 +461,11 @@ const Filters = ({
   useEffect(() => {
     if (Object.keys(filterParameters).length === 0) {
       setFilters(filterParameters);
+      handleClearFilters();
       // Add an empty filter so the user doesn't have to click the 'add filter' button
       generateEmptyFilter();
     }
-  }, [filterParameters, setFilters, generateEmptyFilter]);
+  }, [filterParameters, setFilters, generateEmptyFilter, handleClearFilters]);
```
I thought about refactoring this so that the initial filter state has no filter objects in it instead of a blank placeholder with uuid, but I think that is too big and this works for what we are looking to do. It just makes it a little difficult to rely on the count to show/hide or anything else. I ended up creating some booleans vars that describe what is going on more and removing some checks that weren't actually doing anything but idk.

**Initial state**
```
{
    "id": "79f79105-4718-4df4-a614-fb665ec12b52",
    ...emptyFilterDetailsAsPlaceholder
}
```

**State after completing the first filter**
```
{
    "id": "79f79105-4718-4df4-a614-fb665ec12b52",
    ...completeFilterDetails
}
```

johnclary replied:
yeah, that is tempting for sure. i can see how the initial filter creates lots of headaches. i will capture this tech debt here: https://github.com/cityofaustin/atd-data-tech/issues/10538

mddilley commented on moped-editor/src/components/GridTable/Filters.js:
```diff
@@ -676,38 +745,34 @@ const Filters = ({
           </Button>
         </Grid>
         <Grid item xs={12} md={1}>
-          {Object.keys(filterParameters).length > 0 && (
```
This check wasn't doing anything because the `useEffect` above makes sure that the `filterParameters` object always has at least one key.

chiaberry replied:
🧹  thankyou

mddilley commented on moped-editor/src/components/GridTable/Filters.js:
```diff
@@ -676,38 +745,34 @@ const Filters = ({
           </Grid>
         </Hidden>
         <Grid item xs={12} md={2}>
-          {Object.keys(filterParameters).length > 0 && (
```
Same here, the `Grow` component below has a prop that uses `handleApplyValidation` to control whether it grows in (visible) or not (hidden). The `handleApplyValidation` helper also creates feedback strings about why filters are not valid that we don't display anywhere.

johnclary replied:
nice!

chiaberry commented:
What do you yall think about changing the color of the labels on the radio buttons to the secondary text color, to match the rest of the labels? Is it too dim?
![Screenshot 2023-11-07 at 9 45 20 AM](https://github.com/cityofaustin/atd-moped/assets/12474808/f1b1808d-e53f-4bf8-b585-1d90e01bc767)

johnclary commented on moped-editor/src/components/GridTable/Filters.js:
```diff
@@ -132,20 +133,20 @@ const generateEmptyField = (uuid) => {
    */
   const classes = useStyles();
-  const queryPath = useLocation().pathname;
+  let [, setSearchParams] = useSearchParams();
```
TIL—i did not know you can destructure like this!

chiaberry replied:
same!! I was reading MDN and testing stuff.

johnclary commented on moped-editor/src/views/projects/projectsListView/useProjectListViewQuery/useAdvancedSearch.js:
```diff
@@ -62,12 +57,17 @@ const makeAdvancedSearchWhereString = (filters) =>
+
+  /* Determine or/any from search params if it exists */
+  const isOrFromSearchParams = searchParams.get("isOr");
+  const initialIsOrState = isOrFromSearchParams === "true" ? true : false;
```
✅ i tried to break this feature by setting the url param to `isOr=pizza` and i observed that this was interpreted as the  "match all" setting

johnclary reviewed (changes requested):
@mddilley this is fantastic! 

i have encountered one small issue, which is that the "< All projects" links on the project view does not retain the `isOr` param setting.

<img width="782" alt="Screenshot 2023-11-07 at 2 54 01 PM" src="https://github.com/cityofaustin/atd-moped/assets/14793120/57be5a36-bde9-432c-aea4-dfd8f081619b">

i think that link state is coming from [here](https://github.com/cityofaustin/atd-moped/blob/45fa650bde12579c8663369eb9a2956e4f4e33b2/moped-editor/src/views/projects/projectsListView/ProjectsListViewTable.js#L147). hopefully it's not too hard to get this new param in there.


otherwise, this is working perfectly and it's going to make our users v happy!

chiaberry reviewed (approved):
I left a note about a style tweak, but besides that it looks great! And I am glad you are refactoring the hooks used from react-router-dom. 

I encountered the existing bug that you are working on but couldnt find any other way to break it!

mddilley commented on moped-editor/src/views/projects/projectView/ProjectView.js:
```diff
@@ -175,10 +175,10 @@ const ProjectView = () => {
-  const allProjectsLink = !!previousFilters
-    ? `/moped/projects?filter=${previousFilters}`
-    : "/moped/projects";
+  const { queryString: previousProjectListViewQueryString } = location.state;
+  const allProjectsLink = !previousProjectListViewQueryString
+    ? "/moped/projects"
+    : `/moped/projects${previousProjectListViewQueryString}`;
```
Bigger update here, instead of extending this to pass the state of `isOr` along with the state of `filters`, I updated this to get all query strings using React Router so that this will support any future params too.

johnclary replied:
niiiice!

mddilley commented on moped-editor/src/views/projects/projectsListView/helpers.js:
```diff
@@ -128,8 +128,11 @@ export const useTableComponents = ({
  */
-export const useColumns = ({ hiddenColumns, linkStateFilters, classes }) =>
-  useMemo(
+export const useColumns = ({ hiddenColumns, classes }) => {
+  const location = useLocation();
+  const queryString = location.search;
+
+  const columns = useMemo(
```
This is the source of the `queryString` the comes through the React Router Links that carry state to `ProjectView` so we can link back to the previous filters.

mddilley commented:
@johnclary I added a couple of comments to explain how I fixed the Link that takes users back to the project list view from the project view with their previously applied filters in the url.

johnclary reviewed (approved):
It's fixed! I love seeing how much this code continues to improve. thanks @mddilley!!

mddilley commented:
thanks, y'all! merging 🚢

roseeichelmann commented on moped-editor/src/components/GridTable/Filters.js:
```diff
@@ -367,12 +385,16 @@ const Filters = ({
    * Clears the filters
    */
-  const handleClearFilters = () => {
+  const handleClearFilters = useCallback(() => {
```
i dont think this needs a `useCallback` anymore, in the past this function was called within a `useEffect` but now its not
